pkg/client/vault/client.go

```diff
@@ -0,0 +1,206 @@
+package vault
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/google/uuid"
+	"github.com/hyperledger/aries-framework-go/pkg/crypto"
+	"github.com/hyperledger/aries-framework-go/pkg/crypto/tinkcrypto"
+	"github.com/hyperledger/aries-framework-go/pkg/doc/util/signature"
+	"github.com/hyperledger/aries-framework-go/pkg/kms"
+	"github.com/hyperledger/aries-framework-go/pkg/vdr/fingerprint"
+	"github.com/trustbloc/edge-core/pkg/zcapld"
+	edv "github.com/trustbloc/edv/pkg/client"
+	"github.com/trustbloc/edv/pkg/restapi/models"
+)
+
+const (
+	hubKMSCreateKeyStorePath = "/kms/keystores"
+
+	headerLocation        = "Location"
+	headerXRootCapability = "X-ROOTCAPABILITY"
+)
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Client vault`s client.
+type Client struct {
+	remoteKMSURL string
+	kms          kms.KeyManager
+	crypto       crypto.Crypto
+	edvClient    *edv.Client
+	httpClient   httpClient
+}
+
+// Opt represents Client`s option.
+type Opt func(*Client)
+
+// WithHTTPClient allows providing HTTP client.
+func WithHTTPClient(client httpClient) Opt {
+	return func(vault *Client) {
+		vault.httpClient = client
+	}
+}
+
+// NewClient creates a new vault client.
+func NewClient(kmsURL, edvURL string, kmsClient kms.KeyManager, opts ...Opt) (*Client, error) {
+	cryptoService, err := tinkcrypto.New()
+	if err != nil {
+		return nil, fmt.Errorf("tinkcrypto new: %w", err)
+	}
+
+	client := &Client{
+		remoteKMSURL: kmsURL,
+		kms:          kmsClient,
+		crypto:       cryptoService,
+		edvClient:    edv.New(edvURL),
+		httpClient:   &http.Client{},
+	}
+
+	for _, fn := range opts {
+		fn(client)
+	}
+
+	return client, nil
+}
+
+// CreateVault creates a new vault and KMS store bases on generated DIDKey.
+func (c *Client) CreateVault() (*CreatedVault, error) {
+	didKey, err := c.createDIDKey()
+	if err != nil {
+		return nil, fmt.Errorf("create DID key: %w", err)
+	}
+
+	kmsLoc, err := c.createKeyStore(didKey)
+	if err != nil {
+		return nil, fmt.Errorf("create key store: %w", err)
+	}
+
+	edvLoc, err := c.createDataVault(didKey)
+	if err != nil {
+		return nil, fmt.Errorf("create data vault: %w", err)
+	}
+
+	return &CreatedVault{
+		ID:  uuid.New().String(),
+		KMS: kmsLoc,
+		EDV: edvLoc,
+	}, nil
+}
+
+// CreatedVault represents success response of CreateVault function.
+type CreatedVault struct {
+	ID  string    `json:"id"`
+	EDV *Location `json:"edv"`
+	KMS *Location `json:"kms"`
+}
+
+// Location consists of URI and zcap capability.
+type Location struct {
+	URI  string `json:"uri"`
+	ZCAP string `json:"zcap"`
+}
+
+func (c *Client) createDIDKey() (string, error) {
```

Comment:
Note: the DID method should probably be configurable.

In any case, we won't be using `did:key` in the demo.